File: src/data_rover/core/navigation/evaluate.py
"""Pure navigation evaluator over (Metamodel, Model).

Read-only and session-free: relationship hops go through `model.indexes`
adjacency (never a model scan); property hops instead read the frontier
element's `element.properties` dict directly (an element-reference property
is not indexed as adjacency), so a property hop's cost is per-element, not
per-edge. Scopes filter with the shared search matchers, and type checks use
the Metamodel's cached descendant/ancestor sets — so one hop is O(edges
touching the frontier) or O(1) per property step, matching the read-layer
O(entity) rule.
Like routes/read.py, callers do NOT take the session write_mutex: reads race
benignly against in-memory mutation.

Enumeration is depth-first over SORTED element ids with per-expansion
dedup (parallel edges to the same endpoint yield one continuation), which
makes the chain order deterministic — that determinism is what lets the API
layer do stateless offset/limit paging by re-evaluating.

Two caps bound the work on 80 MB models: `max_visited` counts every edge
examined; `max_chains` bounds the collected output. Hitting either stops
enumeration and flags the result `truncated`.
"""

from __future__ import annotations

from collections.abc import Sequence
from dataclasses import dataclass, field
from typing import TYPE_CHECKING

from data_rover.core.metamodel.schema import Metamodel
from data_rover.core.model.element import Element
from data_rover.core.model.model import Model
from data_rover.core.search.criteria import (
    AnyOfCriterion,
    PropertyCriterion,
    match_element,
)

from .schema import (
    FilterStep,
    NavigationDefinition,
    PathNavigation,
    PropertyStep,
    RelationshipStep,
    RowStart,
    Scope,
    ScriptStep,
    SetExpression,
    StepItem,
)

if TYPE_CHECKING:
    from ..script.embed import ScriptEvalContext


@dataclass(frozen=True)
class EvalLimits:
    max_visited: int = 100_000
    max_chains: int = 5_000


@dataclass(frozen=True)
class PropertyValue:
    """Terminal chain node for a SCALAR property step: when the stepped-on
    property is not element-typed, navigation ends AT the property's value(s)
    instead of hopping to another element. Wrapped (rather than carried raw)
    so a string value can never be mistaken for an element id by downstream
    consumers — every consumer discriminates with ``isinstance(node, str)``.
    Frozen (hashable) because chain nodes are deduped through dict keys."""

    value: str | int | float | bool


#: One position in a chain: an element id, or a terminal scalar value.
ChainNode = str | PropertyValue


@dataclass
class ChainResult:
    """Chains INCLUDE the start element at index 0 (see schema docstring).
    Every node is an element id except a possible trailing `PropertyValue`
    (a scalar property step is always terminal). `warnings` carries script-
    step degradations (pruned chains, dropped ids) generated during THIS
    evaluate call — missing-property prunes stay silent, unchanged."""

    step_types: list[str]
    chains: list[tuple[ChainNode, ...]]
    truncated: bool
    warnings: list[str] = field(default_factory=list)


@dataclass
class _Budget:
    max_visited: int
    visited: int = 0
    exhausted: bool = field(default=False)

    def spend(self, n: int) -> bool:
        """Charge n edge-examinations; False once the budget is gone."""
        self.visited += n
        if self.visited > self.max_visited:
            self.exhausted = True
        return not self.exhausted


def evaluate(
    metamodel: Metamodel,
    model: Model,
    defn: NavigationDefinition,
    limits: EvalLimits = EvalLimits(),
    *,
    row_elements: Sequence[str] | None = None,
    script: ScriptEvalContext | None = None,
) -> ChainResult:
    """Evaluate a ref-free definition (run `resolve_refs` first).

    `row_elements`, when given, binds any `RowStart` sentinel encountered
    (top-level or nested inside a `set_op` operand) to those element ids —
    table columns supply their row's element(s) here. A `RowStart` reached
    with no binding raises `ValueError` (see `RowStart`'s docstring).

    `script`, when given, backs any `ScriptStep` hop encountered (top-level
    or nested); its shared warnings channel is snapshotted at entry so the
    returned `ChainResult.warnings` carries only what THIS call generated."""
    budget = _Budget(max_visited=limits.max_visited)
    w0 = len(script.warnings) if script is not None else 0
    if isinstance(defn, SetExpression):
        members, truncated = _evaluate_set(
            metamodel,
            model,
            defn,
            limits,
            budget,
            row_elements=row_elements,
            script=script,
        )
        return ChainResult(
            step_types=[],
            chains=[(i,) for i in sorted(members)],
            truncated=truncated or budget.exhausted,
            warnings=list(script.warnings[w0:]) if script is not None else [],
        )
    start_ids = _start_ids(
        metamodel, model, defn, limits, budget, row_elements=row_elements, script=script
    )
    chains: list[tuple[ChainNode, ...]] = []
    truncated = False
    for start_id in start_ids:
        if _walk(
            metamodel,
            model,
            defn.steps,
            0,
            (start_id,),
            chains,
            limits,
            budget,
            defn.exclude_visited,
            script,
        ):
            truncated = True
            break
    return ChainResult(
        step_types=[
            s.relationship_type
            if isinstance(s, RelationshipStep)
            else s.property_name
            if isinstance(s, PropertyStep)
            else (s.comment or "script")  # ScriptStep: label per its own docstring
            for s in defn.steps
            if not isinstance(s, FilterStep)
        ],
        chains=chains,
        truncated=truncated or budget.exhausted,
        warnings=list(script.warnings[w0:]) if script is not None else [],
    )


def _start_ids(
    metamodel: Metamodel,
    model: Model,
    defn: PathNavigation,
    limits: EvalLimits,
    budget: _Budget,
    *,
    row_elements: Sequence[str] | None = None,
    script: ScriptEvalContext | None = None,
) -> list[str]:
    if isinstance(defn.start, RowStart):
        if row_elements is None:
            raise ValueError("navigation is row-rooted; no row element bound")
        return sorted(dict.fromkeys(row_elements))
    if isinstance(defn.start, SetExpression):
        members, truncated = _evaluate_set(
            metamodel,
            model,
            defn.start,
            limits,
            budget,
            row_elements=row_elements,
            script=script,
        )
        if truncated:
            budget.exhausted = True
        return sorted(members)
    return _scope_ids(metamodel, model, defn.start)


def _scope_ids(metamodel: Metamodel, model: Model, scope: Scope) -> list[str]:
    if scope.types:
        ids: set[str] = set()
        for type_name in scope.types:
            for concrete in metamodel.element_descendants(type_name):
                ids |= model.indexes.elements_by_type.get(concrete, set())
    else:
        ids = set(model.elements.keys())
    return sorted(i for i in ids if _matches_criteria(model, model.elements[i], scope))


def _match_nav_criterion(model: Model, element: Element, criterion) -> bool:
    """Navigation criterion match. Property criteria are EXISTENCE-GATED: the
    element must actually carry the property (except `exists`/`is_empty`, which
    handle absence explicitly). This intentionally diverges from the shared
    search matcher's coerce-missing-to-'' semantics — `core/search` is
    untouched, so `/model/search` stays byte-identical. An `any_of` group
    recurses HERE (not through the shared matcher) so each member is gated
    exactly as a top-level criterion would be; an empty group stays the
    shared no-op (matches everything)."""
    if isinstance(criterion, AnyOfCriterion):
        return not criterion.criteria or any(
            _match_nav_criterion(model, element, m) for m in criterion.criteria
        )
    if isinstance(criterion, PropertyCriterion) and criterion.op not in (
        "exists",
        "is_empty",
    ):
        if criterion.name not in element.properties:
            return False
    return match_element(model, element, criterion)


def _matches_criteria(model: Model, element: Element, scope: Scope) -> bool:
    return all(_match_nav_criterion(model, element, c) for c in scope.criteria)


def _matches_filter(model: Model, element: Element, step: FilterStep) -> bool:
    return all(_match_nav_criterion(model, element, c) for c in step.criteria)


def _matches_target_types(
    metamodel: Metamodel, element: Element, target_types: list[str]
) -> bool:
    if not target_types:
        return True
    return any(metamodel.is_element_subtype(element.type_name, t) for t in target_types)


def _hop(
    metamodel: Metamodel,
    model: Model,
    element_id: str,
    step: RelationshipStep,
    budget: _Budget,
) -> list[str]:
    idx = model.indexes
    if step.direction == "out":
        rel_ids = set(idx.outgoing_ids(element_id))
    elif step.direction == "in":
        rel_ids = set(idx.incoming_ids(element_id))
    else:  # either — union dedupes self-loops present in both directions
        rel_ids = set(idx.outgoing_ids(element_id)) | set(idx.incoming_ids(element_id))
    if not budget.spend(len(rel_ids)):
        return []
    nxt: set[str] = set()
    for rid in rel_ids:
        rel = model.relationships[rid]
        if not metamodel.is_relationship_subtype(rel.type_name, step.relationship_type):
            continue
        other = rel.target_id if rel.source_id == element_id else rel.source_id
        el = model.elements.get(other)
        if el is not None and _matches_target_types(metamodel, el, step.target_types):
            nxt.add(other)
    return sorted(nxt)


def _hop_property(
    metamodel: Metamodel,
    model: Model,
    element_id: str,
    step: PropertyStep,
    budget: _Budget,
) -> list[ChainNode]:
    """Continuations of a property hop. For an ELEMENT-typed effective property
    def, the element's `property_name` value(s) resolve to existing elements —
    gated on the def's datatype so a string that merely looks like an id must
    not navigate. For a SCALAR property the chain terminates AT the value(s):
    each scalar item becomes a `PropertyValue` terminal (list order preserved —
    it is the stored order, already deterministic), so the value is displayable
    downstream instead of the chain silently vanishing. Absent/unset/dangling
    cases still prune silently (never raise): navigation stays inspectable on
    odd models, mirroring FilterStep's existence-gating. Deliberately NOT
    checked: the RESOLVED element's own type against `prop.datatype` — a
    dangling-conformance id still navigates, since conformance is the
    validation pipeline's job, so the UI's frontier typing (frontierTypesAt)
    may be narrower than the actual frontier."""
    element = model.elements[element_id]
    prop = next(
        (
            p
            for p in metamodel.effective_element_properties(element.type_name)
            if p.name == step.property_name
        ),
        None,
    )
    if prop is None:
        return []
    value = element.properties.get(step.property_name)
    if value is None:
        return []
    candidates = value if isinstance(value, list) else [value]
    if not budget.spend(len(candidates)):
        return []
    if not metamodel.is_element_type(prop.datatype):
        return [
            PropertyValue(item)
            for item in candidates
            if isinstance(item, (str, int, float, bool))
        ]
    resolved: list[ChainNode] = []
    resolved.extend(
        sorted(
            item
            for item in set(candidates)
            if isinstance(item, str) and item in model.elements
        )
    )
    return resolved


def _hop_script(
    model: Model,
    element_id: str,
    step: ScriptStep,
    script: ScriptEvalContext | None,
    budget: _Budget,
) -> list[ChainNode]:
    """Continuations of a script hop: `step(el)` returns the next frontier's
    ids. DEGRADED, NEVER RAISING: a dangling ref, a per-element error, or an
    unknown returned id prunes/drops with a warning on the shared context; an
    unconfigured snippet or absent context prunes silently (mirroring an
    unconfigured navigation source). Dedup preserves the snippet's own return
    order — deterministic because guest output is deterministic."""
    if step.snippet.ref is not None:
        if script is not None:
            script.add_warning(
                f"script step: snippet artifact {step.snippet.ref!r} not found"
            )
        return []
    if step.snippet.definition is None or script is None:
        return []
    res = script.call(step.snippet.definition.code, "step", [element_id])
    if res.error is not None:
        script.add_warning(f"script step failed: {res.error.message}")
        return []
    assert res.value is not None
    raw = list(dict.fromkeys(res.value["ids"]))
    if not budget.spend(len(raw)):
        return []
    known = [i for i in raw if i in model.elements]
    if len(known) != len(raw):
        script.add_warning(
            f"script step returned {len(raw) - len(known)} unknown element id(s)"
        )
    return list(known)


def _walk(
    metamodel: Metamodel,
    model: Model,
    steps: Sequence[StepItem],
    item_idx: int,
    chain: tuple[ChainNode, ...],
    chains: list[tuple[ChainNode, ...]],
    limits: EvalLimits,
    budget: _Budget,
    exclude_visited: bool,
    script: ScriptEvalContext | None = None,
) -> bool:
    """DFS over the interleaved step-item list. A RelationshipStep or
    PropertyStep extends the chain by one hop (deterministic, sorted-id); a
    FilterStep keeps the chain iff its current endpoint matches all criteria,
    adding no column. Returns True when enumeration stopped early (chain cap or
    budget)."""
    if item_idx == len(steps):
        if len(chains) >= limits.max_chains:
            return True
        chains.append(chain)
        return False
    step = steps[item_idx]
    current = chain[-1]
    if not isinstance(current, str):
        # A PropertyValue is TERMINAL — the UI blocks adding steps past a
        # scalar property step, but a hand-written definition may still carry
        # them; such chains prune silently (same stance as absent properties).
        return False
    if isinstance(step, FilterStep):
        if _matches_filter(model, model.elements[current], step):
            return _walk(
                metamodel,
                model,
                steps,
                item_idx + 1,
                chain,
                chains,
                limits,
                budget,
                exclude_visited,
                script,
            )
        return False
    nxt: list[ChainNode]
    if isinstance(step, RelationshipStep):
        nxt = list(_hop(metamodel, model, current, step, budget))
    elif isinstance(step, PropertyStep):
        nxt = _hop_property(metamodel, model, current, step, budget)
    else:  # ScriptStep
        nxt = _hop_script(model, current, step, script, budget)
        if script is not None and exclude_visited:
            # The generic cycle guard below drops silently -- correct for
            # relationship hops (revisits are expected navigation semantics)
            # but a silent mystery for script steps, where an identity return
            # ("keep this element") is the natural idiom. Warn with a count.
            dropped = sum(1 for o in nxt if o in chain)
            if dropped:
                script.add_warning(
                    f"script step: {dropped} element(s) dropped "
                    "(already visited in this chain)"
                )
    if budget.exhausted:
        return True
    for other in nxt:
        if exclude_visited and other in chain:
            continue  # cycle guard: a chain never revisits its own elements
        if _walk(
            metamodel,
            model,
            steps,
            item_idx + 1,
            chain + (other,),
            chains,
            limits,
            budget,
            exclude_visited,
            script,
        ):
            return True
    return False


def _evaluate_set(
    metamodel: Metamodel,
    model: Model,
    expr: SetExpression,
    limits: EvalLimits,
    budget: _Budget,
    *,
    row_elements: Sequence[str] | None = None,
    script: ScriptEvalContext | None = None,
) -> tuple[set[str], bool]:
    """(member ids, any-operand-truncated). `difference` folds left-to-right
    over the operand list; the other ops are order-insensitive.

    Each operand's inner path is evaluated with a FRESH `_Budget` (see
    `_operand_members`) — limits are per-navigation, so one operand hitting
    `max_visited`/`max_chains` never starves a sibling operand's budget.
    Truncation on any operand still propagates into this call's return value
    (and from there into the outer `ChainResult.truncated`).

    `row_elements` passes through to every operand so a `RowStart` nested
    inside a set operand also binds (see `evaluate`). `script` likewise
    passes through so a `ScriptStep` nested inside a set operand's path can
    still hop and report warnings on the shared context."""
    truncated = False
    result: set[str] | None = None
    for operand in expr.operands:
        assert operand.definition is not None  # resolver inlined every ref
        members, op_truncated = _operand_members(
            metamodel,
            model,
            operand.definition,
            operand.step_index,
            limits,
            budget,
            row_elements=row_elements,
            script=script,
        )
        truncated = truncated or op_truncated
        if result is None:
            result = members
        elif expr.op == "union":
            result |= members
        elif expr.op == "intersection":
            result &= members
        elif expr.op == "difference":
            result -= members
        else:  # symmetric_difference
            result ^= members
    return result or set(), truncated


def _operand_members(
    metamodel: Metamodel,
    model: Model,
    defn: NavigationDefinition,
    step_index: int | None,
    limits: EvalLimits,
    budget: _Budget,
    *,
    row_elements: Sequence[str] | None = None,
    script: ScriptEvalContext | None = None,
) -> tuple[set[str], bool]:
    if isinstance(defn, SetExpression):
        # a set has no steps; any explicit index other than 0 is an error
        if step_index not in (None, 0):
            raise ValueError(f"step_index {step_index} out of range for a set operand")
        return _evaluate_set(
            metamodel,
            model,
            defn,
            limits,
            budget,
            row_elements=row_elements,
            script=script,
        )
    inner = evaluate(
        metamodel, model, defn, limits, row_elements=row_elements, script=script
    )
    n_steps = len(inner.step_types)
    index = n_steps if step_index is None else step_index
    if index > n_steps:
        raise ValueError(
            f"step_index {step_index} out of range: path has {n_steps} steps"
        )
    # Set members are ELEMENTS; a PropertyValue terminal at the projected step
    # contributes nothing (a scalar cannot participate in id-set algebra).
    return {
        node for chain in inner.chains if isinstance((node := chain[index]), str)
    }, inner.truncated

File: engine/src/navigation/evaluate.ts
/**
 * The navigation evaluator of `core/navigation/evaluate.py`, line for line,
 * in steps. Chains are enumerated depth first over ids sorted by code point,
 * parallel edges giving one continuation, so the order is the core's and a
 * page can be cut by re-evaluating. Two caps bound the work: `maxVisited`
 * counts every edge and property value examined, `maxChains` the chains
 * collected; either stops the walk and marks the result truncated.
 */
import type { Metamodel } from '../metamodel/metamodel.ts';
import type { Model } from '../model/model.ts';
import { getProp, type ElementRec, type RelRec } from '../model/records.ts';
import {
	compileCriteria,
	matchElement,
	type CompiledCriteria,
	type Criterion
} from '../search/criteria.ts';
import { drain, sortedInSlices, type Progress, type Steps } from '../steps/steps.ts';
import { cmpCodePoint } from '../value/compare.ts';
import { PyFloat, type Value } from '../value/types.ts';
import type {
	FilterStep,
	NavigationDefinition,
	NavigationStep,
	PathNavigation,
	PropertyStep,
	RelationshipStep,
	Scope,
	SetExpression
} from './schema.ts';

export type EvalLimits = { maxVisited: number; maxChains: number };

export const DEFAULT_LIMITS: EvalLimits = { maxVisited: 100_000, maxChains: 5_000 };

/** A value a scalar property step ends its chain at. */
export type ScalarValue = string | number | bigint | PyFloat | boolean;

/**
 * A chain's terminal value, never mistaken for an element id. Its `key` is
 * `(type, value)`: `true`, `1` and `1.0` are three values, as they are three
 * cells.
 */
export class PropertyValue {
	readonly value: ScalarValue;

	constructor(value: ScalarValue) {
		this.value = value;
	}

	get key(): string {
		const { value } = this;
		if (value instanceof PyFloat) return `float:${String(value.value)}`;
		if (typeof value === 'boolean') return `bool:${String(value)}`;
		if (typeof value === 'string') return `str:${value}`;
		return `int:${String(value)}`;
	}
}

/** One position of a chain: an element id, or a terminal value. */
export type ChainNode = string | PropertyValue;

/** Every chain holds its start at index 0; only the last node may be a value. */
export type ChainResult = { stepTypes: string[]; chains: ChainNode[][]; truncated: boolean };

/** The core's `KeyError`: `model.elements[id]` of an id no element has. */
export class NavKeyError extends Error {
	readonly id: string;

	constructor(id: string) {
		super(id);
		this.name = 'NavKeyError';
		this.id = id;
	}
}

/** The core's `ValueError`, its text the core's. */
export class NavValueError extends Error {
	constructor(message: string) {
		super(message);
		this.name = 'NavValueError';
	}
}

// Units of work per step: an edge or a property value examined, an element
// gathered or matched, a set member combined, a chain collected, an id sorted.
const STEP_UNITS = 1024;

/** Counts the units of one call, shared by everything it evaluates, and cuts them into steps. */
export class Meter {
	readonly total: number;
	done = 0;
	private stepStart = 0;

	constructor(total: number) {
		this.total = total;
	}

	/** Counts one unit; true when the step is full and must end. */
	tick(): boolean {
		return ++this.done - this.stepStart >= STEP_UNITS;
	}

	/** Ends the step: what to yield. */
	end(): Progress {
		this.stepStart = this.done;
		return { done: this.done, total: this.total };
	}

	/** Sorts `items` in place or into a new array; use the result. */
	*sort<T>(items: T[], compare: (a: T, b: T) => number): Steps<T[]> {
		if (items.length <= STEP_UNITS) {
			for (let i = 0; i < items.length; i++) if (this.tick()) yield this.end();
			return items.sort(compare);
		}
		if (this.done > this.stepStart) yield this.end();
		const steps = sortedInSlices(items, compare, STEP_UNITS);
		let sorted = 0;
		for (;;) {
			const next = steps.next();
			if (next.done === true) return next.value;
			this.done += next.value.done - sorted;
			sorted = next.value.done;
			yield this.end();
		}
	}
}

/** The edge budget of one path: every operand of a set gets its own. */
class Budget {
	private readonly maxVisited: number;
	private visited = 0;
	exhausted = false;

	constructor(maxVisited: number) {
		this.maxVisited = maxVisited;
	}

	/** Charges `n` examinations; false once the budget is gone. */
	spend(n: number): boolean {
		this.visited += n;
		if (this.visited > this.maxVisited) this.exhausted = true;
		return !this.exhausted;
	}
}

type Context = {
	mm: Metamodel;
	model: Model;
	limits: EvalLimits;
	rowElements: readonly string[] | null;
	meter: Meter;
	compiled: CompiledCriteria;
};

type Walk = Generator<Progress, boolean, void>;

const sortIds = (ctx: Context, ids: string[]) => ctx.meter.sort(ids, cmpCodePoint);

// -- criteria ------------------------------------------------------------------

/**
 * A property criterion needs the property present, bar `exists` and
 * `is_empty`; a group's members are gated one by one, and an empty group
 * matches.
 */
function matchNav(ctx: Context, element: ElementRec, c: Criterion): boolean {
	if (c.type === 'any_of') {
		return c.criteria.length === 0 || c.criteria.some((member) => matchNav(ctx, element, member));
	}
	if (c.type === 'property' && c.op !== 'exists' && c.op !== 'is_empty') {
		if (!Object.hasOwn(element.props, c.name)) return false;
	}
	return matchElement(ctx.model, element, c, ctx.compiled);
}

const matchesAll = (ctx: Context, element: ElementRec, criteria: readonly Criterion[]) =>
	criteria.every((c) => matchNav(ctx, element, c));

/** Every criterion of a definition, set operands included. */
function criteriaOf(defn: NavigationDefinition, into: Criterion[]): Criterion[] {
	const ofSet = (expr: SetExpression) => {
		for (const operand of expr.operands) {
			if (operand.definition !== null) criteriaOf(operand.definition, into);
		}
	};
	if (defn.kind === 'set_op') {
		ofSet(defn);
		return into;
	}
	if (defn.start.kind === 'scope') into.push(...defn.start.criteria);
	else if (defn.start.kind === 'set_op') ofSet(defn.start);
	for (const step of defn.steps) if (step.kind === 'filter') into.push(...step.criteria);
	return into;
}

// -- starts --------------------------------------------------------------------

function* scopeIds(ctx: Context, scope: Scope): Steps<string[]> {
	const { mm, model, meter } = ctx;
	if (scope.types.length > 0) {
		const typed = new Set<string>();
		for (const typeName of scope.types) {
			for (const concrete of mm.elementDescendants(typeName)) {
				for (const element of model.indexes.byType.get(concrete) ?? []) {
					typed.add(element.id);
					if (meter.tick()) yield meter.end();
				}
			}
		}
		if (scope.criteria.length === 0) return yield* sortIds(ctx, [...typed]);
		const matched: string[] = [];
		for (const id of typed) {
			if (matchesAll(ctx, model.getElement(id), scope.criteria)) matched.push(id);
			if (meter.tick()) yield meter.end();
		}
		return yield* sortIds(ctx, matched);
	}
	const ids: string[] = [];
	for (const element of model.elements()) {
		if (scope.criteria.length === 0 || matchesAll(ctx, element, scope.criteria)) {
			ids.push(element.id);
		}
		if (meter.tick()) yield meter.end();
	}
	return yield* sortIds(ctx, ids);
}

function* startIds(ctx: Context, defn: PathNavigation, budget: Budget): Steps<string[]> {
	const { start } = defn;
	if (start.kind === 'row') {
		if (ctx.rowElements === null) {
			throw new NavValueError('navigation is row-rooted; no row element bound');
		}
		return yield* sortIds(ctx, [...new Set(ctx.rowElements)]);
	}
	if (start.kind === 'set_op') {
		const [members, truncated] = yield* evaluateSet(ctx, start, budget);
		if (truncated) budget.exhausted = true;
		return yield* sortIds(ctx, [...members]);
	}
	return yield* scopeIds(ctx, start);
}

// -- hops ----------------------------------------------------------------------

function* hop(
	ctx: Context,
	elementId: string,
	step: RelationshipStep,
	budget: Budget
): Steps<string[]> {
	const { mm, meter } = ctx;
	const element = ctx.model.findElement(elementId);
	// `either` is out plus in, a self-loop, in both, once.
	const out: readonly RelRec[] =
		element === undefined || step.direction === 'in' ? [] : element.out;
	const inward: readonly RelRec[] =
		element === undefined || step.direction === 'out' ? [] : element.in;
	const either = step.direction === 'either';
	let count = out.length;
	for (const rel of inward) if (!either || rel.source !== element) count++;
	if (!budget.spend(count)) return [];
	const next = new Set<string>();
	const visit = (rel: RelRec) => {
		if (!mm.isRelationshipSubtype(rel.typeName, step.relationship_type)) return;
		const other = rel.source.id === elementId ? rel.target : rel.source;
		if (
			step.target_types.length === 0 ||
			step.target_types.some((t) => mm.isElementSubtype(other.typeName, t))
		) {
			next.add(other.id);
		}
	};
	for (const rel of out) {
		visit(rel);
		if (meter.tick()) yield meter.end();
	}
	for (const rel of inward) {
		if (either && rel.source === element) continue;
		visit(rel);
		if (meter.tick()) yield meter.end();
	}
	return yield* sortIds(ctx, [...next]);
}

const isScalar = (item: Value): item is ScalarValue =>
	typeof item === 'string' ||
	typeof item === 'number' ||
	typeof item === 'bigint' ||
	typeof item === 'boolean' ||
	item instanceof PyFloat;

/**
 * Through an element-typed property to the elements its string values name
 * (dangling ones skipped, any other value too); a scalar property ends the
 * chain at each of its values, in list order. A missing definition or value
 * prunes.
 */
function* hopProperty(
	ctx: Context,
	elementId: string,
	step: PropertyStep,
	budget: Budget
): Steps<ChainNode[]> {
	const { mm, model, meter } = ctx;
	const element = model.findElement(elementId);
	if (element === undefined) throw new NavKeyError(elementId);
	const prop = mm
		.effectiveElementProperties(element.typeName)
		.find((p) => p.name === step.property_name);
	if (prop === undefined) return [];
	const value = getProp(element.props, step.property_name);
	if (value === undefined || value === null) return [];
	const candidates = Array.isArray(value) ? value : [value];
	if (!budget.spend(candidates.length)) return [];
	if (!mm.isElementType(prop.datatype)) {
		const values: ChainNode[] = [];
		for (const item of candidates) {
			if (isScalar(item)) values.push(new PropertyValue(item));
			if (meter.tick()) yield meter.end();
		}
		return values;
	}
	const ids = new Set<string>();
	for (const item of candidates) {
		if (typeof item === 'string' && model.findElement(item) !== undefined) ids.add(item);
		if (meter.tick()) yield meter.end();
	}
	return yield* sortIds(ctx, [...ids]);
}

// -- paths ---------------------------------------------------------------------

function stepType(step: Exclude<NavigationStep, FilterStep>): string {
	if (step.kind === 'relationship') return step.relationship_type;
	if (step.kind === 'property') return step.property_name;
	return step.comment || 'script';
}

/**
 * Depth first over the steps from `itemIdx`: true when the walk stopped
 * early, at the chain cap or the budget. A value is terminal, and a script
 * step never runs here — one that reaches a snippet was refused before —
 * so it prunes.
 */
function* walk(
	ctx: Context,
	steps: readonly NavigationStep[],
	itemIdx: number,
	chain: ChainNode[],
	chains: ChainNode[][],
	budget: Budget,
	excludeVisited: boolean
): Walk {
	const { meter } = ctx;
	if (itemIdx === steps.length) {
		if (chains.length >= ctx.limits.maxChains) return true;
		chains.push(chain);
		if (meter.tick()) yield meter.end();
		return false;
	}
	const step = steps[itemIdx]!;
	const current = chain[chain.length - 1]!;
	if (typeof current !== 'string') return false;
	if (step.kind === 'filter') {
		const element = ctx.model.findElement(current);
		if (element === undefined) throw new NavKeyError(current);
		const matched = matchesAll(ctx, element, step.criteria);
		if (meter.tick()) yield meter.end();
		if (!matched) return false;
		return yield* walk(ctx, steps, itemIdx + 1, chain, chains, budget, excludeVisited);
	}
	// The element expanded is a unit too: one with no edges costs a visit.
	if (meter.tick()) yield meter.end();
	let next: readonly ChainNode[] = [];
	if (step.kind === 'relationship') next = yield* hop(ctx, current, step, budget);
	else if (step.kind === 'property') next = yield* hopProperty(ctx, current, step, budget);
	if (budget.exhausted) return true;
	for (const other of next) {
		// A value is never in a chain's prefix: only an id can repeat.
		if (excludeVisited && chain.includes(other)) continue;
		if (yield* walk(ctx, steps, itemIdx + 1, [...chain, other], chains, budget, excludeVisited)) {
			return true;
		}
	}
	return false;
}

// -- sets ----------------------------------------------------------------------

/** The operand's elements at `stepIndex`, and whether its evaluation was cut short. */
function* operandMembers(
	ctx: Context,
	defn: NavigationDefinition,
	stepIndex: number | null,
	budget: Budget
): Steps<[Set<string>, boolean]> {
	if (defn.kind === 'set_op') {
		if (stepIndex !== null && stepIndex !== 0) {
			throw new NavValueError(`step_index ${stepIndex} out of range for a set operand`);
		}
		return yield* evaluateSet(ctx, defn, budget);
	}
	const inner = yield* evaluate(ctx, defn);
	const nSteps = inner.stepTypes.length;
	const index = stepIndex ?? nSteps;
	if (index > nSteps) {
		throw new NavValueError(`step_index ${stepIndex} out of range: path has ${nSteps} steps`);
	}
	const members = new Set<string>();
	for (const chain of inner.chains) {
		const node = chain[index]!;
		if (typeof node === 'string') members.add(node);
		if (ctx.meter.tick()) yield ctx.meter.end();
	}
	return [members, inner.truncated];
}

/** `difference` folds left over the operands; every operand's path has its own budget. */
function* evaluateSet(
	ctx: Context,
	expr: SetExpression,
	budget: Budget
): Steps<[Set<string>, boolean]> {
	const { meter } = ctx;
	let truncated = false;
	let result: Set<string> | null = null;
	for (const operand of expr.operands) {
		const [members, cut] = yield* operandMembers(
			ctx,
			operand.definition!,
			operand.step_index,
			budget
		);
		truncated ||= cut;
		if (result === null) {
			result = members;
			continue;
		}
		if (expr.op === 'intersection') {
			for (const id of result) {
				if (!members.has(id)) result.delete(id);
				if (meter.tick()) yield meter.end();
			}
			continue;
		}
		for (const id of members) {
			if (expr.op === 'union') result.add(id);
			else if (expr.op === 'difference') result.delete(id);
			else if (!result.delete(id)) result.add(id);
			if (meter.tick()) yield meter.end();
		}
	}
	return [result ?? new Set(), truncated];
}

// -- the evaluator ---------------------------------------------------------------

function* evaluate(ctx: Context, defn: NavigationDefinition): Steps<ChainResult> {
	const budget = new Budget(ctx.limits.maxVisited);
	if (defn.kind === 'set_op') {
		const [members, truncated] = yield* evaluateSet(ctx, defn, budget);
		const ids = yield* sortIds(ctx, [...members]);
		return {
			stepTypes: [],
			chains: ids.map((id) => [id]),
			truncated: truncated || budget.exhausted
		};
	}
	const starts = yield* startIds(ctx, defn, budget);
	const chains: ChainNode[][] = [];
	let truncated = false;
	for (const startId of starts) {
		if (yield* walk(ctx, defn.steps, 0, [startId], chains, budget, defn.exclude_visited)) {
			truncated = true;
			break;
		}
	}
	return {
		stepTypes: defn.steps.filter((step) => step.kind !== 'filter').map(stepType),
		chains,
		truncated: truncated || budget.exhausted
	};
}

/**
 * A ref-free definition (see `resolveRefs`) evaluated in steps. `rowElements`
 * binds every row start, nested ones included; a row start with none throws
 * `NavValueError`. Every pattern of the definition is translated before the
 * generator exists, so an unsupported one refuses (501) before any step.
 * `meter` counts the units of the whole call when the caller shares it.
 */
export function evaluateSteps(
	mm: Metamodel,
	model: Model,
	defn: NavigationDefinition,
	limits: EvalLimits = DEFAULT_LIMITS,
	rowElements: readonly string[] | null = null,
	meter: Meter = new Meter(limits.maxVisited)
): Steps<ChainResult> {
	const compiled = compileCriteria(criteriaOf(defn, []));
	return evaluate({ mm, model, limits, rowElements, meter, compiled }, defn);
}

/**
 * The ids `scope` selects in code point order, with no edge budget and no
 * chain cap: a table's scope rows. Its patterns are translated at the first
 * step.
 */
export function* scopeSteps(
	mm: Metamodel,
	model: Model,
	scope: Scope,
	meter: Meter
): Steps<string[]> {
	const compiled = compileCriteria(scope.criteria);
	return yield* scopeIds(
		{ mm, model, limits: DEFAULT_LIMITS, rowElements: null, meter, compiled },
		scope
	);
}

/** `evaluateSteps`, drained. */
export function evaluateNavigationCore(
	mm: Metamodel,
	model: Model,
	defn: NavigationDefinition,
	limits: EvalLimits = DEFAULT_LIMITS,
	rowElements: readonly string[] | null = null
): ChainResult {
	return drain(evaluateSteps(mm, model, defn, limits, rowElements));
}
